src/bin/casr-afl.rs

```diff
@@ -0,0 +1,274 @@
+                .help("Output directory with triaged reports")
+        )
+        .arg(
+            Arg::new("no-cluster")
+                .long("no-cluster")
+                .help("Do not cluster CASR reports")
+        )
+        .get_matches();
+
+    // Init log.
+    let log_level = if matches.value_of("log-level").unwrap() == "debug" {
+        LevelFilter::Debug
+    } else {
+        LevelFilter::Info
+    };
+    let _ = TermLogger::init(
+        log_level,
+        ConfigBuilder::new().set_time_to_local(true).build(),
+        TerminalMode::Mixed,
+    );
+
+    // Get all crashes.
+    let mut crashes: Vec<AflCrashInfo> = Vec::new();
+    for node_dir in fs::read_dir(matches.value_of("input").unwrap())? {
+        let path = node_dir?.path();
+        if !path.is_dir() {
+            continue;
+        }
+
+        // Get crashes from one node.
+        let mut crash_info = AflCrashInfo::default();
+        let cmdline_path = path.join("cmdline");
+        if let Ok(cmdline) = fs::read_to_string(&cmdline_path) {
+            crash_info.target_args = cmdline.split_whitespace().map(|s| s.to_string()).collect();
+            crash_info.is_stdin = !cmdline.contains("@@");
+
+            // Check for ASAN symbols.
+            if let Some(target) = crash_info.target_args.first() {
+                if let Ok(buffer) = fs::read(Path::new(target)) {
+                    if let Ok(elf) = goblin::elf::Elf::parse(&buffer) {
+                        for sym in elf.syms.iter() {
+                            if let Some(name) = elf.strtab.get_at(sym.st_name) {
+                                if name.contains("__asan") {
+                                    crash_info.is_asan = true;
+                                    break;
+                                }
+                            }
+                        }
+                    } else {
+                        error!("Fuzz target: {} must be an ELF executable.", target);
+                        continue;
+                    }
+                } else {
+                    error!("Couldn't read fuzz target binary: {}.", target);
+                    continue;
+                }
+            } else {
+                error!("{} is empty.", cmdline);
+                continue;
+            }
+        } else {
+            error!("Couldn't read {}.", cmdline_path.display());
+            continue;
+        }
+
+        // Push crash paths.
+        for crash in fs::read_dir(path.join("crashes"))? {
+            let crash_path = crash?.path();
+            if crash_path
+                .file_name()
+                .unwrap()
+                .to_str()
+                .unwrap()
+                .starts_with("id:")
+            {
+                crash_info.path = crash_path.to_path_buf();
+                crashes.push(crash_info.clone());
+            }
+        }
+    }
+
+    // Generate CASR reports.
+    info!("Generating CASR reports...");
```

Comment:
Log number of crashes

Reply:
We don't know number of crashes before we start generating reports. Deduplication phase prints reports `before`. Do we need to count crashes while attempting to generate report?

Reply:
Ok, no need